### PR TITLE
Allow user-generated arc line lists

### PR DESCRIPTION
**Pull request by tbowers7**
This PR introduces the ability for users to specify their own line lists to add instrument-specific arc line identifications to a PypeIt reduction, and is primarily in response to #1352 .  This new feature is accomplished via:

- A new command-line script `pypeit_install_linelist` that places a user-specified line list file into the user's PypeIt cache.
- A new function `pypeit.data.get_linelist_filepath()` (paralleling other `get_*_filepath()` functions) that searches both the main code repo and the user's PypeIt cache for the specified file.

Updates to the documentation (`doc/wave_calib.rst`) are included to describe how to use the new functionality (including **big, bold warnings** about using NIST vacuum wavelengths).  Unit tests have been added to cover the new functions.

--------

In the preparation of this PR, several other minor changes were made (ironically making up the bulk of the files touched):

- Finessing the ugly hack in #1425 to clarify the formatting of `reid_arxiv` files.
- Persnickety documentation proofreading, including a reordering of topics in `doc/wave_calib.rst`.
- Moving unused functions from `pypeit/core/wavecal/waveio.py` to `pypeit/deprecated/waveio_old.py`, as discussed with @profxj .
- Adding a flag to `pypeit_identify` allowing for the residuals plot to auto-rescale in search of gross outliers to the fit (especially useful for manual echelle identifications).
- Fixing a nagging error in apparently unexecuted code in `pypeit/core/wavecal/autoid.py`, flagged by VSCode.
- Cleaning up older syntax in `pypeit/core/wavecal/waveio.py` as part of moving functions to `deprecated`.
- Conformal doc file changes.

### CODE REVIEW

**Comment by tbowers7** [comment]
Unit tests pass:
```
$ pytest -Wignore
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
...
collected 200 items                
...                                            
======================= 200 passed in 225.70s (0:03:45) ========================
```

This PR will require a cloud `dev-suite` run.

File: doc/help/pypeit_identify.rst [hunk 1]
```diff
@@ -31,7 +31,9 @@
                             sigma detection for line finding (default: None)
       --pixtol PIXTOL       Pixel tolerance for Auto IDs (default: 0.1)
       --test                Unit tests? (default: False)
-      --linear              Show the spectrum in linear scale? (Default: log
+      --linear              Show the spectrum in linear (rather than log) scale?
```
Comment (profxj) [inline doc/help/pypeit_identify.rst]: nit:  I think we can now remove the `?`

File: doc/wave_calib.rst [hunk 2]
```diff
@@ -44,14 +48,122 @@ be certain by adding the following to the :doc:`pypeit_file`
...
+In the case of the ThAr list, all of the lines are taken from
+the NIST database, and are labelled with a 'MURPHY' flag if the
+line also appears in the list of lines identified by
+`Murphy et al. (2007) MNRAS 378 221 <http://adsabs.harvard.edu/abs/2007MNRAS.378..221M>`_
+
+
+User-Supplied Line Lists
+------------------------
+
+Users may now employ their own line lists, especially those including
```
Comment (profxj) [inline doc/wave_calib.rst]: Somewhere, maybe here, we might add a line like:

`Users who are confident that their new lines would benefit PypeIt broadly, i.e. beyond this single instrument, should consult with the Developers.`

File: pypeit/core/wavecal/waveio.py [hunk 3]
```diff
@@ -188,52 +151,12 @@ def load_line_list(line_file, add_path=False, use_ion=False, NIST=False):
...
-        i1 = line_file.rfind('_')
-        ion = line_file[i0+1:i1]
-        line_list.add_column(Column([ion]*len(line_list), name='Ion', dtype='U5'))
-        line_list.add_column(Column([1]*len(line_list), name='NIST'))
-
-    # Return
-    return line_list
+    line_file = data.get_linelist_filepath(f'{line_file}_lines.dat') if use_ion else \
+        data.get_linelist_filepath(line_file)
+    return astropy.table.Table.read(line_file, format='ascii.fixed_width', comment='#')
```
Comment (profxj) [inline pypeit/core/wavecal/waveio.py:156]: whoa, that's so much cleaner!

File: pypeit/core/wavecal/waveio.py [hunk 4]
```diff
@@ -370,58 +266,6 @@ def load_tree(polygon=4, numsearch=20):
     return file_load, index
 
 
-def load_nist(ion):
```
Comment (profxj) [inline pypeit/core/wavecal/waveio.py:373]: boy, you are touching some of the oldest PypeIt code still around! :)

File: pypeit/tests/test_pkgdata.py [hunk 5]
```diff
@@ -18,32 +21,56 @@ def test_cloud_url():
...
 def test_load_sky_spectrum():
 
-    # Load in the most common sky spectrum to ensure this function works 
-    data.load_sky_spectrum("paranal_sky.fits")
+    # Load in the most common sky spectrum, check that the return is valid
+    skyspec = data.load_sky_spectrum("paranal_sky.fits")
+    assert isinstance(skyspec, xspectrum1d.XSpectrum1D)
+
+
+def test_search_cache():
```
Comment (profxj) [inline pypeit/tests/test_pkgdata.py:64]: thx for all these new tests!

**Review by profxj — APPROVED** [review]
Thx for this new functionality.  A couple of nits from
me but nothing important.

File: doc/wave_calib.rst (hunk 2 above)
Comment (tbowers7) [reply doc/wave_calib.rst]: Added as a `..note`

**Comment by profxj** [comment]
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  200 passed, 1195 warnings in 235.80s (0:03:55) ---
--- PYTEST UNIT TESTS PASSED  115 passed, 1803 warnings in 1165.96s (0:19:25) ---
--- PYTEST VET TESTS FAILED  1 failed, 23 passed, 1551 warnings in 934.18s (0:15:34) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 152/152 TESTS  ---
Testing Started at 2022-07-26T20:24:30.037723
Testing Completed at 2022-07-27T07:11:12.119809
Total Time: 10:46:42.082086

**Comment by profxj** [comment]
`test_identify` failed with this error:

spatid=str(slits.spat_id[slit]), pxtoler=pxtoler,
specname=specname, rescale_resid=rescale_resid)
E       UnboundLocalError: local variable 'wv_calib' referenced before assignment

/PypeIt/pypeit/core/gui/identify.py:351: UnboundLocalError
----------------------------- Captured stdout call -----------------------------

**Comment by tbowers7** [comment]
The `test_identify` VET test now passes:
```
$ pytest vet_tests/test_scripts.py::test_identify -Wignore
============================= test session starts ==============================
platform linux -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
rootdir: /data/tbowers/PypeIt-development-suite
plugins: astropy-header-0.2.1, openfiles-0.5.0, remotedata-0.3.3, cov-3.0.0,
arraydiff-0.5.0, filter-subpackage-0.1.1, mock-3.7.0, doctestplus-0.12.0,
astropy-0.10.0, hypothesis-6.46.11
collected 1 item                                                               

vet_tests/test_scripts.py .                                              [100%]
============================== 1 passed in 5.54s ===============================
```

The fix was needing an `else` clause in `pypeit/core/gui/identify.py`, but this `else` clause has been missing for a long time.  Not sure why the VET test suddenly broke.  It seems related to these lines in `pypeit/scripts/identify.py`:
```
81        # Check if a solution exists
82        solnname = masterframe.construct_file_name(WaveCalib, msarc.master_key,
83                                                   master_dir=msarc.master_dir)
84        # wv_calib = waveio.load_wavelength_calibration(solnname) \
85        wv_calib = WaveCalib.from_file(solnname) \
85            if os.path.exists(solnname) and args.solution else None
```
where `wv_calib` is set to `None`.  This `None` is what is now caught in the new `else` clause added with commit 679784772065eb244ae87b4125ec6f9af6ee04b9.

File: doc/wave_calib.rst [hunk 6]
```diff
@@ -44,14 +48,131 @@ be certain by adding the following to the :doc:`pypeit_file`
...
+PypeIt now allows the use of user-supplied arc line lists for
+wavelength calibration.  The ability to use arbitrary line lists is
+included *caveat emptor*, and users MUST ensure that ALL lists used
+have wavelength measurements **in vacuum**.
+
+.. note::
+
+  Users who are confident that their new lines would benefit PypeIt
+  broadly (*i.e.*, beyond this single instrument) should consult with
+  the Developers.
```
Comment (kbwestfall) [inline doc/wave_calib.rst]: Maybe we should just encourage them to submit a PR?  The "consulting with the Developers" can happen in the PR comments.

File: pypeit/core/wavecal/waveio.py [hunk 7]
```diff
@@ -441,74 +285,17 @@ def load_unknown_list(lines, unknwn_file=None, all=False):
...
-    """
-    # Format
-    tbl['wave'].format = '10.4f'
-    # Write
-    with open(outfile,'w') as f:
-        f.write('# Creation Date: {:s}\n'.format(str(datetime.date.today().strftime('%Y-%m-%d'))))
-        tbl.write(f, format='ascii.fixed_width', overwrite=overwrite)
 
+    # Otherwise
+    msk = np.array([False]*len(line_list))
```
Comment (kbwestfall) [inline pypeit/core/wavecal/waveio.py]: FWIW, you can also do this like so:
```
msk = np.zeros(len(line_list), dtype=bool)
```

**Review by kbwestfall — APPROVED** [review]
Wonderful!  I only had two very minor comments.  Approving now.

File: pypeit/core/wavecal/waveio.py (hunk 7 above)
Comment (tbowers7) [reply pypeit/core/wavecal/waveio.py]: Thanks!  I like the `np.zeros` version better.

File: doc/wave_calib.rst (hunk 6 above)
Comment (tbowers7) [reply doc/wave_calib.rst]: Adding encouragement.

**Pull request merged**